filewatcher/filewatcher.go

```diff
@@ -0,0 +1,159 @@
+	onReload   OnReload
+	logger     logging.Logger
+}
+
+func NewFileWatcher(paths []string, filter loader.Filter, bundleMode bool, store storage.Store, onReload OnReload, logger logging.Logger) *FileWatcher {
+	return &FileWatcher{
+		paths:      paths,
+		filter:     filter,
+		bundleMode: bundleMode,
+		store:      store,
+		onReload:   onReload,
+		logger:     logger,
+	}
+}
+
+func (w *FileWatcher) Start(ctx context.Context) error {
+	watcher, err := w.getWatcher(w.paths)
+	if err != nil {
+		return err
+	}
+	go w.readWatcher(ctx, watcher)
+	return nil
+}
+
+func (w *FileWatcher) getWatcher(rootPaths []string) (*fsnotify.Watcher, error) {
+	watchPaths, err := getWatchPaths(rootPaths)
+	if err != nil {
+		return nil, err
+	}
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, path := range watchPaths {
+		w.logger.WithFields(map[string]interface{}{"path": path}).Debug("watching path")
+		if err := watcher.Add(path); err != nil {
+			return nil, err
+		}
+	}
+
+	return watcher, nil
+}
+
+func (w *FileWatcher) readWatcher(ctx context.Context, watcher *fsnotify.Watcher) {
+	for evt := range watcher.Events {
+		removalMask := fsnotify.Remove | fsnotify.Rename
+		mask := fsnotify.Create | fsnotify.Write | removalMask
+		if (evt.Op & mask) != 0 {
+			w.logger.WithFields(map[string]interface{}{
+				"event": evt.String(),
+			}).Debug("Registered file event.")
+			removed := ""
+			if (evt.Op & removalMask) != 0 {
+				removed = evt.Name
+			}
+			w.processWatcherUpdate(ctx, w.paths, removed)
+		}
+	}
+}
+
+func (w *FileWatcher) processWatcherUpdate(ctx context.Context, paths []string, removed string) {
+	t0 := time.Now()
+
+	loaded, err := initload.LoadPaths(paths, w.filter, w.bundleMode, nil, true, false, nil)
+	if err != nil {
+		w.onReload(ctx, nil, time.Since(t0), w.store, nil, err)
+		return
+	}
+
+	removed = loader.CleanPath(removed)
+
+	err = storage.Txn(ctx, w.store, storage.WriteParams, func(txn storage.Transaction) error {
+		if !w.bundleMode {
+			ids, err := w.store.ListPolicies(ctx, txn)
+			if err != nil {
+				return err
+			}
+			for _, id := range ids {
+				if id == removed {
+					if err := w.store.DeletePolicy(ctx, txn, id); err != nil {
+						return err
+					}
+				} else if _, exists := loaded.Files.Modules[id]; !exists {
+					// This branch get hit in two cases.
+					// 1. Another piece of code has access to the store and inserts
+					//    a policy out-of-band.
+					// 2. In between FS notification and loader.Filtered() call above, a
+					//    policy is removed from disk.
+					bs, err := w.store.GetPolicy(ctx, txn, id)
+					if err != nil {
+						return err
+					}
+					module, err := ast.ParseModule(id, string(bs))
+					if err != nil {
+						return err
+					}
+					loaded.Files.Modules[id] = &loader.RegoFile{
+						Name:   id,
+						Raw:    bs,
+						Parsed: module,
+					}
+				}
+			}
+		}
+
+		// It's up to onReload to update the store with loaded content
```

Comment:
This is my biggest grievance with the current approach. We're doing some updates to the store above, but not _all_ necessary updates. Either, we should be doing all updates, or none. 
The problem with doing all updates here is that currently, some of that, such as bundle activation, is deep within the compile/run of running/testing a policy. This could be refactored, to some extent, but I'm weary of creating even more paths through the code than already exist.